### ISSUE TITLE
Parameters and Gradient is not logged by WandB under FSDP strategy

### ISSUE THREAD

**Issue by weicao1990** [Lightning-AI/lightning#17512]
### Bug description

I find that when using FSDP strategy, the model parameters and gradients are not logged by WandB. However, everything works well if I switch FSDP to native DDP strategy.

Since the gradients are hooked by wandb.run.watch, I am not sure this is a lighting issue or wandb issue.

### What version are you seeing the problem on?

master

### Error messages and logs

```
# Error messages and logs here please
```

### Environment

<details>
  <summary>Current environment</summary>

```
#- Lightning Component (e.g. Trainer, LightningModule, LightningApp, LightningWork, LightningFlow):
#- PyTorch Lightning Version: 2.1.0dev (updated to bleeding-edge)
#- WandB Version: 0.14.0
```

</details>

**Comment by lendle** [comment]
I reported this to wandb a couple of months ago. Here's a workaround I've been using, which I think should also work when not using FSDP:

``
## In your LightningModule, implement
    def on_before_optimizer_step(self, optimizer):
        if self.trainer._logger_connector.should_update_logs:
            with FullyShardedDataParallel.summon_full_params(
                self.get_configured_model(),
                rank0_only=True,
                writeback=False,
                with_grads=True,
            ):
                if self.global_rank == 0:
                    if isinstance(self.logger, WandbLogger):
                        for name, param in self.model.named_parameters():
                            torch_history: TorchHistory = self.logger.experiment._torch
                            torch_history.log_tensor_stats(param, f"parameters/{name}")
                            if param.requires_grad:
                                torch_history.log_tensor_stats(
                                    param.grad, f"gradients/{name}"
                                )

```

**Comment by lendle** [comment]
Bump

**Comment by konstantinjdobler** [comment]
Logging parameters w/ FSDP is working for me without any monkey patching but gradients are not being logged despite using `.watch(..., log="all")`.

I'm calling  `fabric.logger.watch` **after** having called `fabric.setup_module(model)`

**Comment by 42elenz** [comment]
I also can log not gradients with wandb in pl. I am using "ddp_find_unused_parameters_true" as strategy. Is it really due to the strategy that I dont get gradient logs? I am not able to fix it and don't really know what is wrong. Hope someone can help

 ```
if args.wandb:
        name = remove_fold_suffix(name)
        name = f"{name}_{date_tag}"
        wandb_logger_pt = WandbLogger(project=args.project_name, name=name, reinit=True, tags=fold)
if wandb_logger_pt:
        wandb_logger_pt.watch(model, log='all')
    trainer = Trainer(
        max_epochs=args.training.epochs,
        logger=wandb_logger_pt,
        #log_every_n_steps = args.logging.log_every_n_steps,
        callbacks=[checkpoint_callback],
        accelerator="cpu",
        devices= 1,
        precision='16-mixed',
log_every_n_steps=1,
        #strategy="ddp_find_unused_parameters_true"
    )
    trainer.fit(model, train_dataloaders=train_loader, val_dataloaders=val_loader,)
```

**Comment by amorehead** [comment]
@42elenz, I am also not able to see gradients logged when training with FSDP2 using PyTorch Lightning 2.5.1 and WandB 0.19.6. Previously, I was able to see gradients, parameters, and my computational graph logged when training with Lightning's `DDPStrategy` instead. It is interesting that I can still see parameters logged with FSDP2 (just not gradients).

**Comment by KohakuBlueleaf** [comment]
> [@42elenz](https://github.com/42elenz), I am also not able to see gradients logged when training with FSDP2 using PyTorch Lightning 2.5.1 and WandB 0.19.6. Previously, I was able to see gradients, parameters, and my computational graph logged when training with Lightning's `DDPStrategy` instead. Though, it is interesting that I can still see parameters logged with FSDP2 (just not gradients).

I'm using pytorch lightning 2.5.2 + wandb 0.21.0 with pytorch 2.7.1
When I run things with FSDPv1 it also only log the parameter without gradient.
